### DISCUSSION TITLE
How to add shadows to the model

### DISCUSSION THREAD

**Discussion by playblindlythenchangecareers** [sceneview/sceneview#584] [Q&A]
I tried adding shadows to the model as follows, but it had no effect. Is there something wrong with my approach?
`@Composable
fun Dial(dialViewModel: DialViewModel) {
    Box {
        val engine = rememberEngine()
        val view = rememberView(engine)
        val renderer = rememberRenderer(engine)
        val modelLoader = rememberModelLoader(engine)
        val materialLoader = rememberMaterialLoader(engine)
        val environmentLoader = rememberEnvironmentLoader(engine)
        val centerNode = rememberNode(engine)
        val dialRootNode = rememberNode(engine)
        val planeRootNode = rememberNode(engine)
        val pointerRootNode = rememberNode(engine)
        val cameraNode = rememberCameraNode(engine)
        centerNode.addChildNode(cameraNode)
        cameraNode.lookAt(centerNode)
        cameraNode.position = Position(z = 2.9f)
        dialRootNode.addChildNode(planeRootNode)
        dialRootNode.addChildNode(pointerRootNode)
        dialRootNode.position = Position(0f, -0.1f, 0f)
        DialPlane(dialViewModel, modelLoader, planeRootNode)
        DialPointer(dialViewModel, modelLoader, pointerRootNode)
        Scene(
            engine = engine,
            view = view,
            renderer = renderer,
            environmentLoader = environmentLoader,
            modelLoader = modelLoader,
            mainLightNode = DefaultLightNode(engine),
            cameraNode = cameraNode,
            childNodes = listOf(centerNode, dialRootNode),
        )
    }
}

class DefaultLightNode(engine: Engine) : LightNode(
    engine = engine,
    type = LightManager.Type.SUN,
    apply = {
        color(255f, 255f, 0f)
        intensity(100000f)
        direction(0f, 20f, -5.0f)
        castShadows(true)
    })

@Composable
fun DialPointer(dialViewModel: DialViewModel, modelLoader: ModelLoader, pointerRootNode: Node) {
    LogUtils.d("")
    removeChildNode(pointerRootNode, "dialPointer")
    val dialPointer = ModelFactory.createModelNode(modelLoader, dialViewModel.dialPointer).apply {
        isShadowCaster = true
        isShadowReceiver = true
    }
    dialPointer.position = Position(0f, 0f, 0.28f)
    dialPointer.name = "dialPointer"
    pointerRootNode.addChildNode(dialPointer)
}

@Composable
fun DialPlane(dialViewModel: DialViewModel, modelLoader: ModelLoader, planeRootNode: Node) {
    LogUtils.d("")
    removeChildNode(planeRootNode, "dialPlane")
    val dialPlane = ModelFactory.createModelNode(modelLoader, dialViewModel.dialPlane).apply {
        isShadowCaster = true
        isShadowReceiver = true
    }
    dialPlane.name = "dialPlane"
    planeRootNode.addChildNode(dialPlane)
}`

**Comment by ThomasGorisse** [comment]
In v4.0.0-rc.1, shadows come from `LightNode`:

```kotlin
SceneView {
    LightNode(
        apply = {
            intensity(100_000f)
            direction(0f, -1f, -1f)
            castShadows(true)
        }
    )
    ModelNode(modelInstance = model)
}
```

For shadows to appear, you need:
1. A `LightNode` with `castShadows(true)`
2. A receiving surface (e.g. a `CubeNode` as a ground plane, or another model)
3. The model's materials must support shadow casting (most GLB materials do by default)

In `ARSceneView`, the AR session provides default lighting with shadows on detected planes.
